marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs

```diff
@@ -724,35 +751,62 @@ isAddressInTarget (Just targetAddresses) addr =
       C.AddressByron _       -> False
       C.AddressShelley addr' -> addr' `elem` targetAddresses
 
-getUtxos :: (C.IsCardanoEra era) => Maybe TargetAddresses -> C.Tx era -> Map C.TxIn Utxo
-getUtxos maybeTargetAddresses (C.Tx txBody@(C.TxBody C.TxBodyContent {C.txOuts}) _) =
-  fromRight Data.Map.empty (getUtxos' txOuts)
+getTxOutFromTxBodyContent :: C.TxBodyContent build era -> [C.TxOut C.CtxTx era]
+getTxOutFromTxBodyContent C.TxBodyContent {..} = case txScriptValidityToScriptValidity  txScriptValidity of
+  C.ScriptValid   -> txOuts -- When transaction is valid, only transaction fee is collected
+  C.ScriptInvalid -> collateral txReturnCollateral -- failed Tx, we collect from collateral and return excess collateral
```

Comment:
Is `getTxOutFromTxBodyContent` the "main fix" of this PR, i.e the one that is suppose to make the txId/txIx duplication go away? I'm looking at the `getUtxoEvents` function which calls `getUtxos` and `txOutBalanceFromTx` which *both* call `getTxOutFromTxBodyContent` deeper inside -- I wonder if this could be refactored to something simpler that calls `getTxOutFromTxBodyContent` only once? (Because as far as I understand it, the idea is to spend all regular UTxO's when a script is valid and only the collateral if it's not valid. It feels like this should be done once per transaction.)

Reply:
nice catch, thank you, I have a bit of refactoring to do here